enterprise/server/execution_service/execution_service.go

```diff
@@ -73,8 +75,19 @@ func tableExecToProto(in tables.Execution) (*espb.Execution, error) {
 		return nil, err
 	}
 
+	var actionResultDigest *repb.Digest
+	if in.StatusCode == int32(codes.OK) {
+		actionResultDigest = proto.Clone(d).(*repb.Digest)
+	} else {
+		actionResultDigest, err = digest.AddInvocationIDToDigest(d, in.InvocationID)
```

Comment:
Would it make sense to instead directly store the ActionResult proto in the CAS when the action fails (as opposed to fudging the hash and storing it in AC)? Then, I think we wouldn't have to worry about Bazel returning the cached action result of a failed action, since cached actions should only ever be served from AC (IIUC?) and it would avoid the need to adding the invocation ID to the digest. I think we would just need to send the ActionDigest as part of the last Operation in the execution stream, so that the ExecutionServer can store it in the table. + @tylerwilliams

Reply:
I think the current solution is simpler. 
1) The current solution doesn't require us to store additional field in the db. if we store the failed action result in CAS, we would somehow need to store the failed action result digest in the execution table. 
2) The FE code is simpler, fetch the action result from action result digest, if it's not set fall back to action digest. If we store the failed action result in CAS. I would imagine the fetch action result cold would be a bit hairy (if we only store the failed action result in CAS; if we store both successful and failed action result in CAS, then there are duplicate stuff between AC and CAS) 
3) With the current solution, bazel would not be able to get the cached action result of a failed action, even though it's in the AC.

Reply:
+1  (for the reasons lulu mentioned)

I prefer the CL as written to trying to store some ActionResults in the CAS and some in the AC. (Which, btw, would also require munging the digests because the Action is already stored under the digest in the CAS...)

Reply:
Sounds good, thanks for the detailed explanation :)